evm/src/trace/identifier/signatures.rs

```diff
@@ -0,0 +1,135 @@
+use ethers::abi::{Event, Function};
+use foundry_utils::{decode_selector, get_event, get_func, selectors::SelectorType};
+use serde::{Deserialize, Serialize};
+use std::{collections::BTreeMap, io::BufWriter, path::PathBuf};
+use tracing::warn;
+
+/// As identifier that tries to identify functions and events using their signatures.
+#[derive(Debug, Default)]
+pub struct SignaturesIdentifier {
+    cached: CachedSignatures,
+    cached_path: Option<PathBuf>,
+}
+
+impl SignaturesIdentifier {
+    pub fn new(cache_path: Option<PathBuf>) -> eyre::Result<Self> {
+        if let Some(cache_path) = cache_path {
+            let path = cache_path.join("signatures");
+            let cached = if path.is_file() {
+                serde_json::from_reader(std::fs::File::open(&path)?)?
+            } else {
+                if let Err(err) = std::fs::create_dir_all(cache_path) {
+                    warn!(target: "signaturesidentifier", "could not create signatures cache dir: {:?}", err);
+                }
+                CachedSignatures::default()
+            };
+            return Ok(Self { cached, cached_path: Some(path) })
+        }
+        Ok(Self::default())
+    }
+
+    pub fn save(&self) {
+        if let Some(cached_path) = &self.cached_path {
+            if let Ok(file) = std::fs::File::create(&cached_path) {
+                if serde_json::to_writer(BufWriter::new(file), &self.cached).is_err() {
+                    warn!(target: "signaturesidentifier", "could not serialize SignaturesIdentifier");
+                }
+            } else {
+                warn!(target: "signaturesidentifier", "could not open {}", cached_path.to_string_lossy());
+            }
+        }
+    }
+}
+
+impl SignaturesIdentifier {
+    async fn identify<T>(
+        &mut self,
+        selector_type: SelectorType,
+        identifier: &[u8],
+        get_type: fn(&str) -> eyre::Result<T>,
+    ) -> Option<T> {
+        let map = match selector_type {
+            SelectorType::Function => &mut self.cached.functions,
+            SelectorType::Event => &mut self.cached.events,
+        };
+
+        let identifier = format!("0x{}", hex::encode(identifier));
+
+        if !map.contains_key(&identifier) {
+            if let Ok(mut signatures) = decode_selector(&identifier, selector_type).await {
+                if !signatures.is_empty() {
+                    map.insert(identifier.to_string(), signatures.pop().expect(""));
+                }
+            }
+        }
+
+        if let Some(signature) = map.get(&identifier) {
+            return get_type(signature).ok()
```

Comment:
looks like this is the only time the fn is used so I think we should rather return on option instead of taking `get_type` as fn